docs/language/informal/enhanced-type-promotion.md

```diff
@@ -0,0 +1,537 @@
+    condition is false, then `v` is known to be of type T in the else clause.
+
+    ```dart
+    if (o is! String) notHere; else oPromotedToStringHere;
+    ```
+
+### Loop Scopes
+
+Much like if statements, we extend C-style for and while loops to promote inside
+their bodies when the condition is true and the condition expression shows a `v`
+has type T:
+
+```dart
+while (o is String) oPromotedToStringHere;
+
+for (; o is String;;) oPromotedToStringHere;
+```
+
+### Scopes After Exits
+
+The last common case where users expect type promotion to work but currently
+doesn't is in code that can only be reached if a type test succeeds because
+earlier code will cause the code to exit, as in:
+
+```dart
+int stringLength(Object o) {
+  if (o is! String) throw "Not a string.";
+  return o.length; // Should be promoted here.
+}
+```
+
+We say a statement "cannot complete normally" if the statement is one of:
+
+*   A return statement
+*   A rethrow statement
+*   An expression statement whose expression is a throw expression
+*   An if statement with an else where neither statement can complete normally
+*   A block whose last statement cannot complete normally
+
+Then, given an if statement of the form `if (b) s1 else s2` that is contained in
+some block:
+
+*   If `b` shows that `v` has type T when `b` is true, and `s2` cannot complete
+    normally, then `v` has the type T in the statements after the if statement
+    in the immediately enclosing block. As in:
+
+    ```dart
+    int stringLength(Object o) {
+      if (o is String) /* whatever */; else throw "Not a string.";
+      return o.length; // Promoted here.
+    }
+    ```
+
+*   If `b` shows that `v` has type T when `b` is false, and `v` can be promoted,
+    and `s1` cannot complete normally, then `v` has the type T in the statements
+    after the if statement in the immediately enclosing block.
+
+    This is also allowed if the if statement has no else clause at all.
+
+*Optional: We could expand the definition of complete normally for a block to
+allow for returns or throws before the last statement as long as they are
+unconditionally executed.*
+
+## Promoting More Variables
+
+The specification places several restrictions on the variables whose type can be promoted. The variable `v`:
+
+*   Must be either a local variable or a parameter,
+
+*   cannot be potentially assigned to within the scope S that the promotion
+    would apply to,
+
+*   cannot be potentially assigned to within any closure, and
+
+*   if `v` is captured by a closure in the scope S then `v` cannot be
+    potentially mutated anywhere.
+
+This is a little too restrictive. We relax the restriction on potential
+assignments to allow assignments of the same type, as in:
+
+```dart
+int stringLength(Object o) {
+  if (o is String) {
+    o = o.trim();
+    return o.length;
+  }
+}
+```
+
+Here, since `trim()` returns a String, the assignment of `o` to its result does
+not disable its promotion.
```

Comment:
Consider:

```dart
foo(Object o) {
  o.length; // Should have error!
  o = "string now";
  o.length; // This would be OK.
}
```

If we allow assignment promoting to other types, then we need to split the scope at each assignment so that we don't promote until *after* the assignment. That's reasonable, but I think it would make this proposal quite a bit more complex.

I do think we want to get there eventually, but I'm hoping to not do so yet. I'd rather defer the interesting flow-sensitive analysis later and try to do it holistically. This proposal is more about what's the minimal amount we can do shortly so that we can eliminate the ad-hoc type propagation that analyzer is doing?

Is that OK?